python/ray/serve/_private/deployment_state.py

```diff
@@ -1595,55 +1579,76 @@ def autoscale(
-            and self._target_state.adjust_capacity is False
-        ):
+        if decision_num_replicas == self._target_state.target_num_replicas:
             return
 
         logger.info(
             f"Autoscaling replicas for deployment {self.deployment_name} in "
-            f"application {self.app_name} to {clipped_decision_num_replicas}. "
+            f"application {self.app_name} to {decision_num_replicas}. "
             f"current_num_ongoing_requests: {current_num_ongoing_requests}, "
             f"current handle queued queries: {current_handle_queued_queries}."
         )
 
         new_info = copy(self._target_state.info)
-        new_info.set_autoscaled_num_replicas(clipped_decision_num_replicas)
         new_info.version = self._target_state.version.code_version
+
+        has_reached_autoscaling_bounds = self._has_reached_autoscaling_bounds()
+        if has_reached_autoscaling_bounds:
+            new_status_trigger = DeploymentStatusTrigger.AUTOSCALING
+        else:
+            new_status_trigger = self.curr_status_info.status_trigger
+
         self._set_target_state(
             new_info,
-            status_trigger=DeploymentStatusTrigger.AUTOSCALING,
-            adjust_capacity=False,
+            decision_num_replicas,
+            status_trigger=new_status_trigger,
+            allow_scaling_statuses=has_reached_autoscaling_bounds,
+        )
+
+    def _has_reached_autoscaling_bounds(self) -> bool:
+        """Whether or not this deployment has reached the autoscaling bounds.
+
+        This method should only be used for autoscaling deployments. It raises
+        an assertion error otherwise.
+
+        Returns: True if the number of running replicas for the current
+            deployment version has reached a point that's within the
+            autoscaling bounds. Note that if the deployment has already
+            reached such a point and then moves outside the bounds, this
+            method will still return True. False otherwise.
+        """
+
+        target_version = self._target_state.version
+        num_replicas_running_at_target_version = self._replicas.count(
+            states=[ReplicaState.RUNNING], version=target_version
         )
 
+        autoscaling_policy = self._target_state.info.autoscaling_policy
+        assert autoscaling_policy is not None
+
+        lower_bound = autoscaling_policy.get_current_lower_bound(
+            self._target_state.info.target_capacity,
+            self._target_state.info.target_capacity_direction,
+        )
+        upper_bound = get_capacity_adjusted_num_replicas(
+            autoscaling_policy.config.max_replicas,
+            self._target_state.info.target_capacity,
+        )
+        has_reached_autoscaling_bounds = (
+            self.curr_status_info.status_trigger
+            != DeploymentStatusTrigger.CONFIG_UPDATE_STARTED
+            or lower_bound <= num_replicas_running_at_target_version <= upper_bound
+        )
```

Comment:
This makes `_has_reached_autoscaling_bounds` return False if the status trigger is `CONFIG_UPDATE_STARTED`, but the name of the method implies that it's only checking for autoscaling bounds, which is a bit confusing to me. Can this check be moved outside of the method?

Also, can you add a comment explaining why we shouldn't allow scaling statuses if status trigger is config update started?